homeassistant/components/panasonic_viera/media_player.py

```diff
@@ -53,94 +50,97 @@
     | SUPPORT_STOP
 )
 
-PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
-    {
-        vol.Required(CONF_HOST): cv.string,
-        vol.Optional(CONF_MAC): cv.string,
-        vol.Optional(
-            CONF_BROADCAST_ADDRESS, default=DEFAULT_BROADCAST_ADDRESS
-        ): cv.string,
-        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
-        vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
-        vol.Optional(CONF_APP_POWER, default=DEFAULT_APP_POWER): cv.boolean,
-    }
-)
+_LOGGER = logging.getLogger(__name__)
 
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
-    """Set up the Panasonic Viera TV platform."""
-    mac = config.get(CONF_MAC)
-    broadcast = config.get(CONF_BROADCAST_ADDRESS)
-    name = config.get(CONF_NAME)
-    port = config.get(CONF_PORT)
-    app_power = config.get(CONF_APP_POWER)
-
-    if discovery_info:
-        _LOGGER.debug("%s", discovery_info)
-        name = discovery_info.get("name")
-        host = discovery_info.get("host")
-        port = discovery_info.get("port")
-        udn = discovery_info.get("udn")
-        if udn and udn.startswith("uuid:"):
-            uuid = udn[len("uuid:") :]
-        else:
-            uuid = None
-        remote = RemoteControl(host, port)
-        add_entities([PanasonicVieraTVDevice(mac, name, remote, host, app_power, uuid)])
-        return True
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up Panasonic Viera TV from a config entry."""
 
-    host = config.get(CONF_HOST)
-    remote = RemoteControl(host, port)
+    config = config_entry.data
 
-    add_entities(
-        [PanasonicVieraTVDevice(mac, name, remote, host, broadcast, app_power)]
+    host = config[CONF_HOST]
+    mac = config[CONF_MAC]
+    broadcast = config[CONF_BROADCAST_ADDRESS]
+    name = config[CONF_NAME]
+    port = config[CONF_PORT]
+    app_power = config[CONF_APP_POWER]
+
+    uuid = mac if mac else host
+
+    app_id = None
+    encryption_key = None
+
+    remote = None
+
+    try:
+        if CONF_APP_ID in config and CONF_ENCRYPTION_KEY in config:
+            app_id = config[CONF_APP_ID]
+            encryption_key = config[CONF_ENCRYPTION_KEY]
+            remote = RemoteControl(
+                host, port, app_id=app_id, encryption_key=encryption_key
+            )
+        else:
+            remote = RemoteControl(host, port)
```

Comment:
Since this does I/O, we need to run that in the executor thread pool with `hass.async_add_executor_job`. And to be able to pass keyword arguments we need to use `functools.partial`.
```suggestion
            remote = await hass.async_add_executor_job(
                partial(RemoteControl, host, port, **params)
            )
```